### PR TITLE
Catch `ServerDownOrUnreachableError` in redfish_available function

### PR DESCRIPTION
**Pull request by chanchiwai-ray**
See PR title.

Closes: #93

### CODE REVIEW

**Review by jneo8 — CHANGES_REQUESTED** [review]
In view of the particularity of this charm.

Instead of just catch one more `ServerDownOrUnreachableError`, we should provide a general exception catch to prevent any possible error to crash the install hook.

In this case any error happen when detecting redfish, then we just leave a error msg and disable the redfish.

How do you think?

(Intentionally make install hook fail is also a way, but we should discuss about this.)

**Comment by chanchiwai-ray** [comment]
we can't catch general exception: `W0718: Catching too general exception Exception (broad-exception-caught)`

**Comment by Pjack** [comment]
I would like to confirm some hehavior of hardware-observer-operator.

1. If Redfish is temporaray unavailable (e.g. network / hardware issue), will the redfish collector be disabled?
2. After collector is been disabled, when the redfish connection recovered, will the collector be enabled again?
3. What's the behavior of alert?

**Comment by chanchiwai-ray** [comment]
Ideally, the redfish collector should be temporarily disabled, then when redfish service is recovered, the collector should be automatically enabled. But looking at the code for the charm's life-cycle, in the case you mentioned, it will:

1. exporter will enable the collector `on_config_changed`
2. but the tool will not be installed if it's not installed at the begining.

**Comment by jneo8** [comment]
> we can't catch general exception: W0718: Catching too general exception Exception (broad-exception-caught)

It just for linter check. As long as we know what we are doing, we can choose to ignore it.

File: src/hw_tools.py [hunk 1]
```diff
@@ -354,11 +359,15 @@ def redfish_available() -> bool:
...
-    except RetriesExhaustedError:  # redfish not available
+    except (RetriesExhaustedError, ServerDownOrUnreachableError):
+        # redfish not available
         result = False
     except (SessionCreationError, InvalidCredentialsError):
         # redfish available, wrong credentials or not able to create a session
         result = True
+    except Exception:  # pylint: disable=W0718
+        # mark redfish unavailable for any generic exception
+        result = False
```
Comment (jneo8) [inline src/hw_tools.py:370]: I would like to provide some logs here.

**Review by jneo8 — CHANGES_REQUESTED** [review]
Overall LGTM. Only need 1 more line change.

File: src/hw_tools.py (hunk 1 above)
Comment (chanchiwai-ray) [reply src/hw_tools.py:370]: Added

**Comment by Pjack** [comment]
> Ideally, the redfish collector should be temporarily disabled, then when redfish service is recovered, the collector should be automatically enabled. But looking at the code for the charm's life-cycle, in the case you mentioned, it will:
> 
> 0. tool will not be installed and collector will not be enabled (because of `get_hw_tool_white_list`)
> 1. exporter will enable the collector `on_config_changed`
> 2. but the tool will not be installed if it's not installed at the beginning.

The tool will not be installed => the collector cannot collect data even when it is enabled. 
I think this is a bug. If we do not fix it in this MR, could we raise another issue to track this issue?
Just like you mentioned, In idea case, the collector and the tool should be automatically enabled when the hardware back to normal. 
> 
> Note: this is generic, not specific to redfish

**Comment by chanchiwai-ray** [comment]
No problem, I opened #114 to track that issue

File: src/hw_tools.py (hunk 1 above)
Comment (jneo8) [reply src/hw_tools.py:370]: The level should be warning or error instead of info.

**Review by jneo8 — CHANGES_REQUESTED** [review]
The logging level need to be either warning or error.

File: src/hw_tools.py (hunk 1 above)
Comment (chanchiwai-ray) [reply src/hw_tools.py:370]: Done, it's because other part of the codes are using `info` when exception happens. I wanted to be consistent.

**Comment by jneo8** [comment]
Wait for failed test pass.

**Review by jneo8 — APPROVED** [review]

**Pull request merged**